grails-app/services/org/pih/warehouse/putaway/PutawayService.groovy

```diff
@@ -88,18 +78,36 @@ class PutawayService {
     }
 
     List<PutawayItem> getPendingItems(Location location) {
-        List<Order> orders = Order.findAllByOriginAndOrderType(location, OrderType.findByCode(Constants.PUTAWAY_ORDER))
-        List<Putaway> putaways = orders.collect { Putaway.createFromOrder(it) }
-        List<PutawayItem> putawayItems = []
-
-        putaways.each {
-            putawayItems.addAll(it.putawayItems.findAll {
-                it.putawayStatus == PutawayStatus.PENDING ||
-                        (it.putawayStatus == PutawayStatus.CANCELED && it.splitItems?.any { item -> item.putawayStatus == PutawayStatus.PENDING })
-            })
+        List<Order> orders = Order.createCriteria().list {
+            setResultTransformer(CriteriaSpecification.DISTINCT_ROOT_ENTITY)
+            and {
+                eq("origin", location)
+                eq("orderType", OrderType.findByCode(Constants.PUTAWAY_ORDER))
+            }
+            createAlias("origin", "o")
+            createAlias("picklist", "p", JoinType.LEFT_OUTER_JOIN)
+            createAlias("orderItems", "oi", JoinType.LEFT_OUTER_JOIN)
+            createAlias("oi.product", "oip", JoinType.LEFT_OUTER_JOIN)
+            createAlias("oi.inventoryItem", "ii",JoinType.LEFT_OUTER_JOIN)
+            createAlias("oi.recipient", "oir", JoinType.LEFT_OUTER_JOIN)
+            createAlias("oip.category", "oipc", JoinType.LEFT_OUTER_JOIN)
+            createAlias("oip.synonyms", "oips",JoinType.LEFT_OUTER_JOIN)
+            createAlias("oi.originBinLocation", "oibl", JoinType.LEFT_OUTER_JOIN)
+            createAlias("oi.destinationBinLocation", "oidl", JoinType.LEFT_OUTER_JOIN)
+            createAlias("oi.orderItems", "oioi", JoinType.LEFT_OUTER_JOIN)
```

Comment:
I'm assuming this is pre-fetching the relationships in batches so that they're not lazy fetched one by one when creating the putaway item. Is this the recommended approach? A google search shows there's also these options:

`Book.list(fetch:[authors:"eager"])`

```
A.withCriteria() {
    fetchmode "cs", FetchMode.JOIN
    fetchMode "cs.b", FetchMode.JOIN
}
```
I'm not sure the differences behind the scenes but we should probably determine which of these eager fetching approaches we want to stick with for future refactors

Reply:
> Is this the recommended approach?
I would say there is no big difference in those approaches since they generate the same sql statement that joins the tables and pre-fetch the relationships so they do not create n+1 problem when accessing them in a loop.

In terms of the approach, I would say that the context is important - if you don't need any filtering logic, you could potentially use the `Book.list(fetch:[authors:"eager"])`, but if you need the filtering, I would say the `.createCriteria` + `createAlias` is the most readable and has been used by us since the Grails 3 migration and the lazy intialization exception problems we faced. 
As for the `fetchMode`, we have it in a few places, and I assume they work, but surprisingly when I tried it some time ago, it didn't seem to work, so probably worth to investigate when there would be an occasion, but I would recommend the `createAlias` anyway.
One more possible approach is a data service with the `@Join` annotation (only possible for one association) or `@Query` annotation with the HQL query that joins assocations (you can check it out, we have it in a few places).